synapse/util/caches/response_cache.py

```diff
@@ -75,75 +90,82 @@ def __init__(self, clock: Clock, name: str, timeout_ms: float = 0):
         self._metrics = register_cache("response_cache", name, self, resizable=False)
 
     def size(self) -> int:
-        return len(self.pending_result_cache)
+        return len(self._result_cache)
 
     def __len__(self) -> int:
         return self.size()
 
-    def get(self, key: KV) -> Optional[defer.Deferred]:
-        """Look up the given key.
+    def keys(self) -> Iterable[KV]:
+        """Get the keys currently in the result cache
 
-        Returns a new Deferred (which also doesn't follow the synapse
-        logcontext rules). You will probably want to make_deferred_yieldable the result.
+        Returns both incomplete entries, and (if the timeout on this cache is non-zero),
+        complete entries which are still in the cache.
+
+        Note that the returned iterator is not safe in the face of concurrent execution:
+        behaviour is undefined if `wrap` is called during iteration.
+        """
+        return self._result_cache.keys()
 
-        If there is no entry for the key, returns None.
+    def _get(self, key: KV) -> Optional[ResponseCacheEntry]:
+        """Look up the given key.
 
         Args:
-            key: key to get/set in the cache
+            key: key to get in the cache
 
         Returns:
-            None if there is no entry for this key; otherwise a deferred which
-            resolves to the result.
+            The entry for this key, if any; else None.
         """
-        result = self.pending_result_cache.get(key)
-        if result is not None:
+        entry = self._result_cache.get(key)
+        if entry is not None:
             self._metrics.inc_hits()
-            return result.observe()
+            return entry
         else:
             self._metrics.inc_misses()
             return None
 
     def _set(
-        self, context: ResponseCacheContext[KV], deferred: "defer.Deferred[RV]"
-    ) -> "defer.Deferred[RV]":
+        self,
+        context: ResponseCacheContext[KV],
+        deferred: "defer.Deferred[RV]",
+        opentracing_span_context: Any,
+    ) -> ResponseCacheEntry:
         """Set the entry for the given key to the given deferred.
 
         *deferred* should run its callbacks in the sentinel logcontext (ie,
         you should wrap normal synapse deferreds with
         synapse.logging.context.run_in_background).
 
-        Returns a new Deferred (which also doesn't follow the synapse logcontext rules).
-        You will probably want to make_deferred_yieldable the result.
```

Comment:
Was this changed at some point in the past to no longer be true? I don't think I see changes in here related to this?

Reply:
no, it was true until this PR. see https://github.com/matrix-org/synapse/pull/11607/files#diff-4e303a929ca8d0022d41884af5d1dd65997f3c21e596ee09b66d806144770b25R168 which changes the result from a `Deferred` to a `ResponseCacheEntry`.

Reply:
I was not specific enough -- is it useful to somewhere say something about the synapse logcontext rules still or is that no longer needed?

Reply:
well, it's no longer relevant in `_set`, at least.  ObservableDeferred has its own set of guidance on the logcontext rules (see https://github.com/matrix-org/synapse/blob/d4033b2be6eae2eda9fd5791f7efd80f396963d2/synapse/util/async_helpers.py#L83-L86) - though that does make me realise I should probably add some notes about it on `AbstractObservableDeferred`.

Reply:
That was a bit what I was wondering. I did notice that we call `make_deferred_yieldable` internally. 👍

Reply:
right, done in 6a89434.